packages/test/end-to-end-tests/src/test/documentDirty.spec.ts

```diff
@@ -284,98 +249,104 @@ describe("Document Dirty", () => {
-                "Document will have been marked clean twice more");
-        });
+                // Disconnect the client.
+                documentServiceFactory.disconnectClient(container.clientId, "Disconnected for testing");
 
-        it(`sets ops while connected, but disconnects before sending ops,
-        then reconnects to process them`, async () => {
-            // Set batch values in DDSes in disconnected state.
-            dataObject.context.containerRuntime.orderSequentially(() => {
-                sharedMap.set("key1", "value1");
-                sharedMap.set("key2", "value2");
-            });
+                // State not affect after disconnect
+                checkDirtyState("after disconnect with value set", true, 0);
 
-            assert.equal(wasMarkedDirtyCount, 1,
-                `Document will have incremented the dirty count due to the value set`);
+                // Wait for the Container to get reconnected.
+                await waitForContainerReconnection(container);
 
-            assert.equal(containerRuntime.isDocumentDirty(), true,
-                "Document is marked dirty on edit");
+                // Document should still be dirty right after reconnection
+                checkDirtyState("after reconnect and replayed ops", true, 0);
 
-            assert(container.clientId);
+                // Wait for the ops to get processed.
+                await opProcessingController.process();
 
-            // Disconnect the client.
-            documentServiceFactory.disconnectClient(container.clientId, "Disconnected for testing");
+                // Document will have been marked clean after process
+                checkDirtyState("after processing replayed ops", false, 1);
+            });
+    });
+
+    describe("Force readonly", () => {
+        it(`sets operations when force readonly and then turn off force readonly to process them`, async () => {
+            container.forceReadonly(true);
+            await ensureForContainerConnected(container);
 
-            assert.equal(wasMarkedDirtyCount, 1,
-                `Document will not increment the dirty count as it was already dirty`);
+            // Set values in DDSes in disconnected state.
+            sharedMap.set("key", "value");
 
-            assert.equal(containerRuntime.isDocumentDirty(), true,
-                "Document is still dirty on disconnect");
+            await opProcessingController.process();
 
-            // Wait for the Container to get reconnected.
-            await waitForContainerReconnection(container);
+            // Document should have been marked dirty again due to pending DDS ops
+            checkDirtyState("after value set while force readonly", true, 0);
 
-            // Document will have been marked clean on reconnection
-            assert.equal(wasMarkedCleanCount, 1,
-                "Document will have been marked clean on reconnection");
+            container.forceReadonly(false);
```

Comment:
not sure if you check anywhere, if not, can you please check that container is actually in read-only state after flipping true and is not after false?
The best way to validate it is to have a component that reports last seen "readonly" event on datastore runtime (I think)

Reply:
@vladris has a PR for some forceReadonly specific test.  I will leave it to be added there.